libs/cardano-ledger-core/src/Cardano/Ledger/BaseTypes.hs

```diff
@@ -328,53 +335,42 @@ instance
   where
   decCBOR = do
     r <- decodeRationalWithTag
-    case fromRationalBoundedRatio r of
+    case boundRational r of
       Nothing ->
         cborError $ DecoderErrorCustom "BoundedRatio" (Text.pack $ show r)
       Just u -> pure u
 
-instance ToJSON (BoundedRatio b Word64) where
-  toJSON = toJSON . toScientificBoundedRatioWord64WithRounding
-
-toScientificBoundedRatioWord64WithRounding :: BoundedRatio b Word64 -> Scientific
-toScientificBoundedRatioWord64WithRounding (toRationalBoundedRatio -> ur) =
-  scientific q 0 + scientific ((r * scale) `quot` d) (negate exp10)
-  where
-    n = numerator ur
-    d = denominator ur
-    (q, r) = n `quotRem` d
-    -- We need to reduce precision for numbers bigger than 1 in order to make them
-    -- parsable without overflowing
-    exp10 = 19 - min 19 (numDigits q)
-    scale = 10 ^ exp10
-    numDigits :: Integer -> Int
-    numDigits = go 0
-      where
-        go ds 0 = ds
-        go ds i = ds `seq` go (ds + 1) (i `quot` 10)
+instance Bounded (BoundedRatio b Word64) => ToJSON (BoundedRatio b Word64) where
+  toJSON = toRationalJSON . unboundRational
 
 instance Bounded (BoundedRatio b Word64) => FromJSON (BoundedRatio b Word64) where
-  parseJSON = either fail pure . fromScientificBoundedRatioWord64 <=< parseJSON
+  parseJSON = \case
+    rational@(Object _) -> parseWith fromRationalEither rational
+    sci -> parseWith fromScientificBoundedRatioWord64 sci
+    where
+      parseWith f = either fail pure . f <=< parseJSON
 
 fromScientificBoundedRatioWord64 ::
   Bounded (BoundedRatio b Word64) =>
   Scientific ->
   Either String (BoundedRatio b Word64)
 fromScientificBoundedRatioWord64 (normalize -> sci)
-  | coeff < 0 = failWith "negative"
+  | coeff < 0 = failWith "negative" sci
   | exp10 <= 0 = do
-      when (exp10 < -19) $ failWith "too precise"
+      when (exp10 < -19) $ failWith "too precise" sci
       fromRationalEither (coeff % (10 ^ negate exp10))
   | otherwise = do
-      when (19 < exp10) $ failWith "too big"
+      when (19 < exp10) $ failWith "too big" sci
```

Comment:
`19` is the number of decimal digits for 64 bit values. I think it might be a good idea to consider removing this magic number by introducing a constant. Something along the lines of `numberOfDecimals64 = 19`. Would that be desirable? If so, I can do it here and now or later as part of some refactoring.